### PR TITLE
8269851: OperatingSystemMXBean getProcessCpuLoad reports incorrect process cpu usage in containers

### PR DESCRIPTION
**Pull request by xpbob**
…ocess cpu usage in containers

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269851](https://bugs.openjdk.java.net/browse/JDK-8269851): OperatingSystemMXBean getProcessCpuLoad reports incorrect process cpu usage in containers

### Reviewers
 * [Severin Gehwolf](https://openjdk.java.net/census#sgehwolf) (@jerboaa - **Reviewer**)

### Contributors
 * Severin Gehwolf `<sgehwolf@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4702/head:pull/4702` \
`$ git checkout pull/4702`

Update a local copy of the PR: \
`$ git checkout pull/4702` \
`$ git pull https://git.openjdk.java.net/jdk pull/4702/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4702`

View PR using the GUI difftool: \
`$ git pr show -t 4702`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4702.diff">https://git.openjdk.java.net/jdk/pull/4702.diff</a>

</details>

### CODE REVIEW

**Comment by xpbob** [comment]
getProcessCpuLoad return jvm cpu load。
8 core in machine, set 1 core for container。
Run same program, the getProcessCpuLoad return the same value。

expected
same program run in machine,value 0.12
run in 1 core container,value 0.99
ProcessCpuLoad is associative container setting

**Comment by dholmes-ora** [comment]
Why did you create a new PR ??

**Comment by xpbob** [comment]
> Why did you create a new PR ??

bug description changed。
update the commit info.

File: src/jdk.management/unix/classes/com/sun/management/internal/OperatingSystemImpl.java
Comment (linzang) [inline src/jdk.management/unix/classes/com/sun/management/internal/OperatingSystemImpl.java:49]: I think this empty line could be removed

**Comment by linzang** [comment]
Hi Bob,
     I am not a reviewer, just a tiny comments that you may need to remove the unnecessary empty line. Thanks.
-Lin

**Review by YaSuenag — COMMENTED** [review]
In HotSpot, CPU usage for JVM would be calculated at `get_jvm_tickes()` in os_perf_linux.cpp . It refers `/proc/self/stat` - should we refer it in same way?

I discussed about similar issue in #4299 , then we came to an agreement that we should refactor that HotSpot and JMX should share the code. Of course you can fix JMX only in this PR, but it still uses different implementation from HotSpot. I prefer to fix it in code consolidation between HotSpot and JMX.

**Review by jerboaa — CHANGES_REQUESTED** [review]
This patch seems fine in principle, but I don't like the duplication. Right now, `getProcessUsageDividesTotal()` and `getUsageDividesTotal()` look almost the same, but use different fields to set. Perhaps pass a parameter whether or not it's a system/process load request and use the different setters accordingly. Same duplication between `getCpuLoad()` and `getProcessCpuLoad()`. The main difference seems to be how `usageNanos` are being retrieved.

**Comment by jerboaa** [comment]
> In HotSpot, CPU usage for JVM would be calculated at `get_jvm_tickes()` in os_perf_linux.cpp . It refers `/proc/self/stat` - should we refer it in same way?
> 
> I discussed about similar issue in #4299 , then we came to an agreement that we should refactor that HotSpot and JMX should share the code. Of course you can fix JMX only in this PR, but it still uses different implementation from HotSpot. I prefer to fix it in code consolidation between HotSpot and JMX.

Yes, we need a bug to track this work (or at least decide on a plan of action - if any). I wouldn't conflate this into this bug, though as it's not clear whether or not JFR/thread dumps should report container values. I'll try to create a bug for this consolidation effort today.

**Comment by jerboaa** [comment]
> This patch seems fine in principle [...]

Let me revise this. It looks like imposing CPU limits via cpuset-cpus would go down a path which won't have this fixed with the current patch.

**Comment by xpbob** [comment]
> > This patch seems fine in principle [...]
> 
> Let me revise this. It looks like imposing CPU limits via cpuset-cpus would go down a path which won't have this fixed with the current patch.

use (getProcessCpuLoad0() * totalCPUs / containerCPUs) for CPU limits via cpuset-cpus

**Review by jerboaa — CHANGES_REQUESTED** [review]
Thanks for the update. `getProcessCpuLoad()` and `getCpuLoad()` looks way too similar to me. Is there a way to make them more generic and do the process vs. system paths as needed? I think there is.

**Comment by xpbob** [comment]
> Thanks for the update. `getProcessCpuLoad()` and `getCpuLoad()` looks way too similar to me. Is there a way to make them more generic and do the process vs. system paths as needed? I think there is.

use getCpuLoadWithTarget for process and system

**Comment by jerboaa** [comment]
> > Thanks for the update. `getProcessCpuLoad()` and `getCpuLoad()` looks way too similar to me. Is there a way to make them more generic and do the process vs. system paths as needed? I think there is.
> 
> use getCpuLoadWithTarget for process and system

Thanks. We are getting there. I've proposed a PR to your branch which makes this a bit easier to read:
https://github.com/xpbob/jdk/pull/1

If you merge that it'll show up in this PR.

**Comment by xpbob** [comment]
> > > Thanks for the update. `getProcessCpuLoad()` and `getCpuLoad()` looks way too similar to me. Is there a way to make them more generic and do the process vs. system paths as needed? I think there is.
> > 
> > 
> > use getCpuLoadWithTarget for process and system
> 
> Thanks. We are getting there. I've proposed a PR to your branch which makes this a bit easier to read:
> [xpbob#1](https://github.com/xpbob/jdk/pull/1)
> 
> If you merge that it'll show up in this PR.

Thank you so much @jerboaa .

Code has been merged.
I tested the following cgroup scenario on Linux/x64 and all passed.
```
  cpu-quota=50000 cpu-period=50000 
  cpuset-cpus=0-2
  cpu.shares=2048
  cpuset-cpus=3
```

**Comment by jerboaa** [comment]
https://github.com/jerboaa/jdk/actions/runs/1026610314 is a pre-integration run of this.

File: src/jdk.management/unix/classes/com/sun/management/internal/OperatingSystemImpl.java
Comment (jerboaa) [inline src/jdk.management/unix/classes/com/sun/management/internal/OperatingSystemImpl.java:98]: The old code had these comments here. We should add them as they're still relevant:

```
                // If CPU quotas and shares are not active then find the average system load for
                // all online CPUs that are allowed to run this container.

                // If the cpuset is the same as the host's one there is no need to iterate over each CPU
```

File: src/jdk.management/unix/classes/com/sun/management/internal/OperatingSystemImpl.java
Comment (jerboaa) [inline src/jdk.management/unix/classes/com/sun/management/internal/OperatingSystemImpl.java:106]: Please add the comment as it was in the original code:

```
                        // cgroups is mounted, but CPU resource is not limited.
                        // We can assume the VM is run on the host CPUs.
```

**Review by jerboaa — APPROVED** [review]
Except for the missing comments in the new code blocks this looks good to me.

**Review by jerboaa — APPROVED** [review]
Still looks good.

**Comment by xpbob** [comment]
May I get a second review for this change?  Thanks.

**Comment by DamonFool** [comment]
Will sponsor it tomorrow if there is no objection.
Thanks.

**Pull request closed without merging**
